c++/src/Writer.cc

```diff
@@ -0,0 +1,659 @@
+
+    void add(ColumnVectorBatch& rowsToAdd) override;
+
+    void close() override;
+
+  private:
+    void Init();
+    void InitStripe();
+    void WriteStripe();
+    void WriteMetadata();
+    void WriteFileFooter();
+    void WritePostscript();
+    void BuildFooterType(const Type& t, proto::Footer& footer, uint32_t& index);
+    static proto::CompressionKind convertCompressionKind(
+                                                  const CompressionKind& kind);
+  };
+
+  const char * WriterImpl::magicId = "ORC";
+
+  WriterImpl::WriterImpl(
+                         const Type& t,
+                         OutputStream * stream,
+                         const WriterOptions& opts) :
+                         outStream(stream),
+                         options(opts),
+                         type(t) {
+    streamsFactory = createStreamsFactory(options, outStream);
+    columnWriter = buildWriter(type, *streamsFactory, options);
+    stripeRows = totalRows = indexRows = 0;
+    currentOffset = 0;
+
+    uint64_t bufferCapacity = 4 * 1024 * 1024; // 4M
+    compressionStream = createCompressor(
+                                  options.getCompression(),
+                                  outStream,
+                                  options.getCompressionStrategy(),
+                                  bufferCapacity,
+                                  options.getBlockSize(),
+                                  *options.getMemoryPool());
+    bufferedStream.reset(new BufferedOutputStream(
+                                            *options.getMemoryPool(),
+                                            outStream,
+                                            bufferCapacity,
+                                            options.getBlockSize()));
+
+    Init();
+  }
+
+  std::unique_ptr<ColumnVectorBatch> WriterImpl::createRowBatch(uint64_t size)
+                                                                         const {
+    return type.createRowBatch(size, *options.getMemoryPool());
+  }
+
+  void WriterImpl::add(ColumnVectorBatch& rowsToAdd) {
+    if (options.getEnableIndex()) {
+      uint64_t pos = 0;
+      uint64_t chunkSize = 0;
+      uint64_t rowIndexStride = options.getRowIndexStride();
+      while (pos < rowsToAdd.numElements) {
+        chunkSize = std::min(rowsToAdd.numElements - pos,
+                             rowIndexStride - indexRows);
+        columnWriter->add(rowsToAdd, pos, chunkSize);
+
+        pos += chunkSize;
+        indexRows += chunkSize;
+        stripeRows += chunkSize;
+
+        if (indexRows >= rowIndexStride) {
+          columnWriter->createRowIndexEntry();
+          indexRows = 0;
+        }
+      }
+    } else {
+      stripeRows += rowsToAdd.numElements;
+      columnWriter->add(rowsToAdd, 0, rowsToAdd.numElements);
+    }
+
+    if (columnWriter->getEstimatedSize() >= options.getStripeSize()) {
+      WriteStripe();
+    }
```

Comment:
We won't throw exception here. If estimated size is smaller than stripe size, it just means the current data added are not big enough to form a stripe. We continue keeping data in memory until a subsequent add() that satisfies this check, or a call to Writer::close() which forces us to flush any remaining data as the last stripe of file.